### PR TITLE
feat_: the router - add candidates by taking the max amount on enabled chains

### PR DESCRIPTION
**Pull request by saledjenic**
In this PR, we are addressing improvements logged in this issue: https://github.com/status-im/status-desktop/issues/15395

Also balance map is created at the beginning of the suggested routes function and shared with other functions when needed.

### CODE REVIEW

File: services/wallet/router/router_v2.go [hunk 1]
```diff
@@ -492,40 +490,225 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+		lockedAmount, fromChainLocked := input.FromLockedAmount[selectedFromChain.ChainID]
+		if fromChainLocked {
+			amountToSend = lockedAmount.ToInt()
+			amountLocked = true
+		} else if len(input.FromLockedAmount) > 0 {
+			for chainID, lockedAmount := range input.FromLockedAmount {
+				if chainID == selectedFromChain.ChainID {
+					continue
+				}
+				amountToSend = new(big.Int).Sub(amountToSend, lockedAmount.ToInt())
```
Comment (alaibe) [inline services/wallet/router/router_v2.go:620]: if amount lock is last chain in the loop, how do you capture correct setup for other chain?
It is possible that last chain modify amount to send, and then you don't capture other possibilities

File: services/wallet/router/router_v2.go [hunk 2]
```diff
@@ -492,40 +490,225 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+
+			if amountLocked {
+				continue
+			}
+
+			amountToSplitAccrossChains := new(big.Int).Set(amountToSend)
 
+			// If the amount that need to be send is bigger than the balance on the chain, then we want to check options if that
+			// amount can be splitted and sent across multiple chains.
+			if input.SendType == Transfer && len(selectedFromChains) > 1 {
```
Comment (alaibe) [inline services/wallet/router/router_v2.go:639]: what about bridge? e.g do a send mainnet to mainnet and optimism to mainnet

**Review by alaibe — COMMENTED** [review]
Would that produce correct result?
Scenario A
Send 20 eth.
Balance mainnet 20 eth, cost 10 dollar
balance optimism 12 eth, cost 1 dollar
balance arbitrum 12 eth, cost 1 dollar
it should produce optimism to optimism
and arbitrum to arbitrum

Scenario B
Balance mainnet 20 eth, cost 10 dollar
balance optimism 12 eth, cost 1 dollar for bridge
balance arbitrum 12 eth, cost 1 dollar for bridge

But want to receive only on mainnet
it should produce optimism to mainnet
and arbitrum to mainnet

File: services/wallet/router/router_v2.go (hunk 1 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go:620]: It doesn't matter if it's the last or not, what we're trying to do here is to calculate an amount that can be sent from this chain if this chain is not among locked chains.

For example, it means:
The user is trying to send 1 ETH, but he locked only 0.2 for optimism and there are no disabled chains (all chains are available).
Now that means that an amount we have to send through one or more chains, where none of them is the optimism chain, is 0.8ETH and that's what we calculate here.

File: services/wallet/router/router_v2.go (hunk 2 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go:639]: I didn't add it, as the Bridge send type per se, cause when you're bridging you usually want to move your token from one network to another, the amount you have on that network. But not sure now when you're asking do we need it, it will cover this type of cases:

user has:
mainnet - 1 eth
opt - 1 eth
arb - 1 eth

and wants to bridge 2 eth to optimism, will result in no routes, cause without adding `Bridge` type to that condition (which is the case now) that's not possible, but if we add it, it will result in having 2 bridges:
mainnet -> opt
arb -> opt

But just to clarify, the above-mentioned bridges, actually all possibilities (transfer+bridge), are taken into consideration for Transfer sending type.

**Comment by saledjenic** [comment]
> Would that produce correct result?
Scenario A
Send 10 eth.
Balance mainnet 20 eth, cost 10 dollar
balance optimism 12 eth, cost 1 dollar
balance arbitrum 12 eth, cost 1 dollar
it should produce optimism to optimism
and arbitrum to arbitrum

It will find 3 options for sending amount:
mainent -> mainnet
opt -> opt
arb -> arb

and then, the resolve candidates function will create possible candidates for sending amount in those 3 cases and after that resolveRoutes function will find the cheapest one and return a single one as the best route. 


> Scenario B
Balance mainnet 20 eth, cost 10 dollar
balance optimism 12 eth, cost 1 dollar for bridge
balance arbitrum 12 eth, cost 1 dollar for bridge

In this scenario if you're still sending 10eth, and if you have opt and arb chains disabled on "to" side, then you will get the same amount options as a result of the `findOptionsForSendingAmount` function call, but when we process those amounts in `resolveCandidates` then you will get 3 possible candidates:
mainent -> mainnet
opt -> opt
arb -> arb
and `resolveRoutes` will return the best one.

But for example if you have mainnet disabled on the "from" side, and still have opt and arb chains disabled on "to" side, then `findOptionsForSendingAmount` will give you:
opt -> opt
arb -> arb
as options for sending amounts, and after that `resolveCandidates` will give you 2 possible candidates, and after that `resolveRoutes` will return a cheaper one as the best route.

----

So to summarize, `findOptionsForSendingAmount` function is not about candidates or the cheapest route, it's about preparing possible options, in which way we can distribute the amount across chains and then `resolveCandidates` function will make the list of candidates for those options, and after that `resolveRoutes` will find the cheapest route.

File: services/wallet/router/router_v2_test.go [hunk 3]
```diff
@@ -2600,420 +2852,441 @@ func TestNoBalanceForTheBestRouteRouterV2(t *testing.T) {
...
-			} else {
-				assert.EqualError(t, err, tc.expectedError.Error())
-			}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			selectedFromChains, _, err := router.getSelectedChains(tt.input)
+			assert.NoError(t, err)
+
+			amountOptions, err := router.findOptionsForSendingAmount(context.Background(), tt.input, selectedFromChains, tt.input.testParams.balanceMap)
```
Comment (alaibe) [inline services/wallet/router/router_v2_test.go]: testing implementation details?

Comment (saledjenic) [reply services/wallet/router/router_v2_test.go]: yes, we have tests for it

**Review by alaibe — APPROVED** [review]

File: services/wallet/router/router_v2.go [hunk 4]
```diff
@@ -492,40 +490,225 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+	balanceMap = make(map[string]*big.Int)
+
+	for _, chain := range selectedFromChains {
+		token := input.SendType.FindToken(r.tokenManager, r.collectiblesService, input.AddrFrom, chain, input.TokenID)
+		if token == nil {
+			continue
+		}
+
+		// add token balance for the chain
+		tokenBalance := big.NewInt(1)
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go:522]: Why token balance is 1 if it is not a collectible?

File: services/wallet/router/router_v2.go [hunk 5]
```diff
@@ -492,40 +490,225 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+
+func (r *Router) findOptionsForSendingAmount(ctx context.Context, input *RouteInputParams, selectedFromChains []*params.Network,
+	balanceMap map[string]*big.Int) (map[uint64][]amountOption, error) {
+	// All we do in this block we're free to do, because of the validateInputData function which checks if the locked amount
+	// was properly set and if there is something unexpected it will return an error and we will not reach this point
+	finalCrossChainAmountOptions := make(map[uint64][]amountOption) // represents all possible amounts that can be sent from the "from" chain
+
+	for _, selectedFromChain := range selectedFromChains {
+
+		amountLocked := false
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go:609]: `amountOption := getAmountOptionForChain(selectedFromChain.ChainID, input)`

to make it smaller

File: services/wallet/router/router_v2.go [hunk 6]
```diff
@@ -492,40 +490,225 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+			}
+
+			amountToSplitAccrossChains := new(big.Int).Set(amountToSend)
 
+			// If the amount that need to be send is bigger than the balance on the chain, then we want to check options if that
+			// amount can be splitted and sent across multiple chains.
+			if input.SendType == Transfer && len(selectedFromChains) > 1 {
+				// All we do in this block we're free to do, because of the validateInputData function which checks if the locked amount
+				// was properly set and if there is something unexpected it will return an error and we will not reach this point
+				selectedButNotLockedChains := []*params.Network{selectedFromChain} // always add the processing chain at the beginning
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go]: `selectedButNotLockedChains := getSelectedUnlockedChains(selectedFromChains, exceptChain = selectedFromChain.ChainID, input)`

Just to make the func smaller

File: services/wallet/router/router_v2.go [hunk 7]
```diff
@@ -492,40 +490,225 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+				for _, net := range selectedFromChains {
+					if net.ChainID == selectedFromChain.ChainID {
+						continue
+					}
+					if _, ok := input.FromLockedAmount[net.ChainID]; !ok {
+						selectedButNotLockedChains = append(selectedButNotLockedChains, net)
+					}
+				}
+
+				crossChainAmountOptions := make(map[uint64][]amountOption)
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go]: `crossChainAmountOptions := getCrossChainAmountOptions(selectedButNotLockedChains, balanceMap)`

File: services/wallet/router/router_v2.go [hunk 8]
```diff
@@ -487,225 +485,391 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+								amount: tokenBalance,
+								locked: false,
+							})
+							amountToSend = new(big.Int).Sub(amountToSend, tokenBalance)
+						} else if amountToSend.Cmp(pathprocessor.ZeroBigIntValue) > 0 {
+							crossChainAmountOptions[chain.ChainID] = append(crossChainAmountOptions[chain.ChainID], amountOption{
+								amount: amountToSend,
+								locked: false,
+							})
+							amountToSend = big.NewInt(0)
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go]: Should we have after it

if amountToSend.Cmp(pathprocessor.ZeroBigIntValue) == 0 {
    break
}

Otherwise next iteration tokenBalance can be 0 and amountToSend == 0 as well and we will add a 0 value option.

File: services/wallet/router/router_v2.go [hunk 9]
```diff
@@ -487,225 +485,391 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+				}
+			}
+		}
+	}
 
-					gasLimit, err := pProcessor.EstimateGas(processorInputParams)
-					if err != nil {
-						continue
-					}
+	// filter out duplicates values for the same chain
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go:672]: separate function?

File: services/wallet/router/router_v2.go [hunk 10]
```diff
@@ -812,7 +952,7 @@ func (r *Router) resolveRoutes(ctx context.Context, input *RouteInputParams, can
 	for len(allRoutes) > 0 {
 		best := findBestV2(allRoutes, tokenPrice, nativeChainTokenPrice)
 
-		err := r.checkBalancesForTheBestRoute(ctx, best, input)
+		err := r.checkBalancesForTheBestRoute(ctx, best, input, balanceMap)
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go:976]: Just curious if we have already filtered out all the routes at the resolving candidates stage, when some paths did not have enough balance, why we need to check it again for the best route?
Should best route be calculated already with respect to balances?

**Review by IvanBelyakoff — COMMENTED** [review]
Sale, good job, I have minor comments inline, some of them to reduce the big function into pieces (they already often have your comments above the corresponding blocks)

File: services/wallet/router/router_v2.go (hunk 4 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go:522]: That code is kept from the old router, it is in the case of ERC721.

File: services/wallet/router/router_v2.go (hunk 8 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go]: since we're in the loop, I think we should break here, since amountToSend is 0

File: services/wallet/router/router_v2.go (hunk 10 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go:976]: We have an issue for that https://github.com/status-im/status-desktop/issues/15397 that will be an improvement after we establish an error reporting mechanism. 
We plan to skip checking if an amount can be sent via a single transfer on a single network if we detect that there is not enough balance on that network, for now, that check is done at the end of the algorithm, but even when we do that, searching for the best route will be done at the end, we cannot do that at the moment of resolving candidates.

**Comment by saledjenic** [comment]
@IvanBelyakoff logic split into smaller functions.

**Review by IvanBelyakoff — APPROVED** [review]
